test/mb_debug_optimus.py

```diff
@@ -0,0 +1,667 @@
+        print(f'num_last_seg: {self.num_last_seg}')
+        print(f'num_sa: {self.num_sa}')
+        print(f'num_fail: {self.num_fail}')
+        print(f'num_dupl: {self.num_dupl}')
+        print(f'num_supl: {self.num_supl}')
+# EPY: END code
+
+# EPY: START code
+op_multiple_primary = 0
+cr_multiple_primary = 0
+
+for idx in discordant_flag_indices:
+    op_flags = [attr.flag for attr in op_master_rec_attr_list[idx]]
+    cr_flags = [attr.flag for attr in cr_master_rec_attr_list[idx]]
+    
+    # reads with more than one primary alignment (!)
+    # according to SAM specs, each read should have only one primary alignment
+    # Note: the issue stems from the synthetic data -- will follow up; for now, let us
+    # ignore such reads
+    op_recs_multiple_primary = sum(flag & MASK_PRIMARY == 0 for flag in op_flags) > 1
+    cr_recs_multiple_primary = sum(flag & MASK_PRIMARY == 0 for flag in cr_flags) > 1
+    op_multiple_primary += op_recs_multiple_primary
+    cr_multiple_primary += cr_recs_multiple_primary
+# EPY: END code
+
+# EPY: START code
+op_flag_summary_stats = FlagSummaryStatistics()
+cr_flag_summary_stats = FlagSummaryStatistics()
+
+for idx in discordant_flag_indices:
+    op_flags = [attr.flag for attr in op_master_rec_attr_list[idx]]
+    cr_flags = [attr.flag for attr in cr_master_rec_attr_list[idx]]
+
+    op_flag_summary_stats.process_single(op_flags, primary_only=False)
+    cr_flag_summary_stats.process_single(cr_flags, primary_only=False)
+# EPY: END code
+
+# EPY: START code
+op_primary_only_flag_summary_stats = FlagSummaryStatistics()
+cr_primary_only_flag_summary_stats = FlagSummaryStatistics()
+
+for idx in discordant_flag_indices:
+    op_flags = [attr.flag for attr in op_master_rec_attr_list[idx]]
+    cr_flags = [attr.flag for attr in cr_master_rec_attr_list[idx]]
+
+    op_primary_only_flag_summary_stats.process_single(op_flags, primary_only=True)
+    cr_primary_only_flag_summary_stats.process_single(cr_flags, primary_only=True)
+# EPY: END code
+
+# EPY: START code
+primary_only_flag_xor_summary_stats = FlagSummaryStatistics()
+primary_only_flag_and_summary_stats = FlagSummaryStatistics()
+
+for idx in discordant_flag_indices:
+    op_flags = [attr.flag for attr in op_master_rec_attr_list[idx]]
+    cr_flags = [attr.flag for attr in cr_master_rec_attr_list[idx]]
+
+    primary_only_flag_xor_summary_stats.process_difference(
+        op_flags, cr_flags, primary_only=True, joint_op='xor')
+
+    primary_only_flag_and_summary_stats.process_difference(
+        op_flags, cr_flags, primary_only=True, joint_op='and')
+# EPY: END code
+
+# EPY: START code
+print(f'Number of Optimus reads with multiple primary alignments: {op_multiple_primary}, fraction = {op_multiple_primary/num_total_records:.4f}')
+print(f'Number of CellRanger reads with multiple primary alignments: {cr_multiple_primary}, fraction = {cr_multiple_primary/num_total_records:.4f}')
+print()
+
+print('Optimus (primary only):')
```

Comment:
In addition to your observations, I think it's interesting that Cell Ranger seems to make approximately twice as many reads as Optimus as duplicates. 

Related, I believe primary alignments are being marked randomly in cases where two reads have identical quality; a different random seed could explain this discrepancy in primary reads.

Reply:
The RC difference is a bit perplexing, since the alignments themselves are identical by name and num positions and were aligned to the same reference. Could that be a difference in STAR version? A bug somewhere?

Reply:
Looking at the numbers of primary, secondary, and duplicate alignments, cell ranger must mark some primary alignments as duplicates, which is not something that Picard does, I believe. 

Evidence: same number of secondary alignments between Optimus and Cell Ranger, but higher numbers of duplicates. 

If I had to guess, Cell Ranger might be marking  any replicated read as a duplicate, whereas Optimus does not mark primary reads that are duplicated as duplicates.